doc/dev/backend-plugins.md

```diff
@@ -0,0 +1,271 @@
+
+- The exact `parentKey` that was returned by the previous phase (`poststorage`, `precommit` or `commit`) of this `set` operation.
+  The key name and value of this key are read-only.
+- The exact keyset `ks` that was returned by the `poststorage` phase of this `set` operation (which is the same one that was returned by the `prestorage` phase).
+  All keys in this keyset and the keyset itself are fully read-only.
+
+There are no restrictions on the `precommit` phase, other than those enforced by `parentKey` and `ks` being (partially) read-only.
+This phase can be used for file-level manipulations, like file-based encryption, line ending conversion or adding file signatures.
+In this sense, it is the counter-part of the `prestorage` phase of the `set` operation.
+
+In the `commit` phase the backend plugin:
+
+- **MUST** make the changes done during the `storage` phase of this `set` operation permanent in such a way that a following `get` operation will be able to read them (assuming there is no other `set` operation in between).
+
+There are no restrictions on what the backend plugin may do in the `postcommit` phase.
+However, it is important to keep in mind that an error in the `postcommit` phase will **not** make the `set` operation fail.
+Once the `commit` phase completes successfully, the `set` operation is also deemed successful, since the changes were made permanent.
+If an error does occur in the `postcommit` phase, it is reported as warning.
+This makes the `postcommit` phase mostly useful for logging.
+
+Finally, `libelektra-kdb` merges the keyset returned by the `postcommit` phase (which is still the same one that was returned by the `prestorage` phase) with the ones returned by other backend plugins for different mountpoints and then returns it to the user.
+
+TODO: Are modifications to `parentKey` visible to the user?
+
+#### Rollback Phases (`set` only)
+
+If any of the phases `prestorage`, `storage`, `poststorage`, `precommit` or `commit` fail, `libelektra-kdb` will continue with the rollback phases.
+Even though this phase is part of the `set` operation, `libelektra-kdb` calls `elektra<Plugin>Error` and not `elektra<Plugin>Set` for these phases.
+
+Similar to the commit phases, the rollback phases (`prerollback`, `rollback` and `postrollback`) are called with:
+
+- The exact `parentKey` that was returned by the phase of this `set` operation that reported an error.
+  The key name and value of this key are read-only.
+- The exact keyset `ks` that was returned by the phase of this `set` operation that reported an error.
+  All keys in this keyset and the keyset itself are fully read-only.
+
+Additionally, the phase that reported an error is communicated to the backend plugin via the global keyset (together with the current phase).
+The value of the key `system:/elektra/kdb/backend/failedphase` is set to the failed phase.
+
+The `prerollback` and `postrollback` phases are mostly useful for logging.
```

Comment:
Do we actually need them?

Reply:
In this first version I tried to keep as close as possible to the current version. In particular, I kept all current positions, because I haven't yet checked whether there are any useful plugins that use them.

Reply:
I just saw that postrollback is also needed as hook for shell scripts, so yes, let us keep the positions.